Log Splits to Range Event Table

mrtracy opened the pull request:
First commit:  _Refactor TestServer client helpers_
Refactor a few helper functions for connecting to a TestServer instance through
a client. This primarily extracts the Postgres URL generation function from the
SQL package and places it into a new TestUtils package for more broad usage.
This method is somewhat complicated, as making a secure postgres driver
connection involves the creation of temporary on-disk certificates.

Commit also adds a small convenience method to open a KVDB connection to the
TestServer, which is somewhat less complex than opening a Postgres connection.

Second Commit: _Begin logging splits to Range Event Log_
This commit begins the logging of split events in the recently added range event
log. This is the first event being logged, and thus some additional
infrastructure was also added.

Creates a new "InternalExecutor" object in the SQL package, which can be used by
other cockroach packages to execute SQL without needing to open a SQL
connection. This needs to be a persistent object because a LeaseManager is
required. The InternalExecutor used by each store currently shares a
LeaseManager with the Executor used by that store.

Commit also extends the rangelog.event schema to hold information relevant to
splits. This information should not necessarily be considered complete;
additional information may be captured in subsequent commits.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3687)

<!-- Reviewable:end -->

Commit:
Refactor TestServer client helpers

Refactor a few helper functions for connecting to a TestServer instance through
a client. This primarily extracts the Postgres URL generation function from the
SQL package and places it into a new TestUtils package for more broad usage.
This method is somewhat complicated, as making a secure postgres driver
connection involves the creation of temporary on-disk certificates.

Commit also adds a small convenience method to open a KVDB connection to the
TestServer, which is somewhat less complex than opening a Postgres connection.

Commit:
Begin logging splits to Range Event Log

This commit begins the logging of split events in the recently added range event
log. This is the first event being logged, and thus some additional
infrastructure was also added.

Creates a new "InternalExecutor" object in the SQL package, which can be used by
other cockroach packages to execute SQL without needing to open a SQL
connection. This needs to be a persistent object because a LeaseManager is
required. The InternalExecutor used by each store currently shares a
LeaseManager with the Executor used by that store.

Commit also extends the rangelog.event schema to hold information relevant to
splits. This information should not necessarily be considered complete;
additional information may be captured in subsequent commits.

mrtracy commented:
Review status: 0 of 15 files reviewed at latest revision, 1 unresolved discussion.

---

*[storage/log.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TL8_YvYnioQ-ZnAJf:-K7TL8_YvYnioQ-ZnAJg:-847703915) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L49)):*
otherRangeID is nullable because other event types (such as a replica change) will not have a second rangeID.  These event types are not yet implemented, but I wanted to work out how to handle nullable fields up front.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Reviewed 8 of 8 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[sql/internal.go, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tud6TcCcfZMjuFSX6:-K7Tud6TcCcfZMjuFSX7:1672578488) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/sql/internal.go#L1)):*
2016 (elsewhere too probably)

---

*[sql/internal.go, line 24 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TunEoTnVn3dyfet7v:-K7TunEoTnVn3dyfet7w:1403038088) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/sql/internal.go#L24)):*
the word "processes" is an awkward choice here

---

*[sql/internal.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TuZOAD3ivw6z6qrd-:-K7TuZOB-t_vRH3KNll9:1938893776) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/sql/internal.go#L34)):*
why pointer receiver?

---

*[storage/log.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TvpDk85_PUzyNsuMZ:-K7TvpDmH9HpxSgXl6rH:1964012856) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L33)):*
consider using a normal iota-based enum here with the mapping to strings implemented using `stringer`. this current approach has the disadvantage that string consts can be silently coerced to `RangeEventLogType`s http://play.golang.org/p/3byBnM7zlL

---

*[storage/log.go, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TwMXDkZYyH66IZoSC:-K7TwMXEHx-bPGdzFRZu:1345074973) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L40)):*
double space after the period.

also s/describes/defines/

---

*[storage/log.go, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TweB9YwtfxB1V9ulj:-K7TweB9YwtfxB1V9ulk:-920799761) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L71)):*
how about:
```
args := [...]interface{}{
	event.timestamp,
	event.rangeID,
	event.eventType,
	event.storeID,
	nil,
}
if event.otherRangeID != nil {
	args[4] = *event.otherRangeID
}
```
the comments aren't helpful, since they exactly mirror these field names on `event`

---

*[storage/log.go, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Txvp8HVwxeARRIyQk:-K7Txvp8HVwxeARRIyQl:781520360) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L85)):*
s/effected/affected/ and also you're not checking that it returned zero, you're checking that it returned non-one. best to include actual number in this message.

---

*[storage/log.go, line 100 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Ty7B_pu0qrH6p46l0:-K7Ty7B_pu0qrH6p46l1:-248617207) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L100)):*
i believe we call these left and right-hand sides of the split elsewhere, but perhaps this is clearer.

---

*[storage/log_test.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TyQjdcftnbrvSzAP8:-K7TyQjeVo6kc0fgdMdL:2015969207) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L48)):*
you can and should shadow using `:=` here. also, does this need a comment about the placeholder bug?

---

*[storage/log_test.go, line 77 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tyj_maPsk7ffy_2rZ:-K7Tyj_maPsk7ffy_2r_:-1715625420) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L77)):*
ditto re placeholder bug comment

---

*[storage/log_test.go, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tz5YMWGqioAXB5PKh:-K7Tz5YMWGqioAXB5PKi:-1971905478) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L82)):*
you only need to check this once after `rows.Next()` stops returning `true`

---

*[storage/log_test.go, line 88 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TywR1H3ISekX7Bckc:-K7TywR1H3ISekX7Bckd:636881036) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L88)):*
nit: inline into the previous row?

---

*[storage/log_test.go, line 96 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tyz2t0PsmIWbC4dLQ:-K7Tyz2t0PsmIWbC4dLR:1399768557) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L96)):*
s/I D/ID /

---

*[testutils/sqlutils/conn.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtZZ27ZFG9w72iQqJ:-K7TtZZ27ZFG9w72iQqK:-5590960) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/testutils/sqlutils/conn.go#L33)):*
not quite right; the function must be called after the connection created from this string is no longer needed.

---

*[testutils/sqlutils/conn.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtQFA9sLnhA2H4Xrq:-K7TtQFA9sLnhA2H4Xrr:-1648980990) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/testutils/sqlutils/conn.go#L40)):*
nit: how come `ts *server.TestServer, t util.Tester` are reversed compared to the signature of this function in its old location?

---

*[util/testing.go, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Ttr7qK-d2j_NtOFhS:-K7Ttr7qK-d2j_NtOFhT:366117756) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/util/testing.go#L69)):*
a function

---

*[util/testing.go, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtvbmdBUi-tKCQWyr:-K7TtvbmdBUi-tKCQWys:510931928) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/util/testing.go#L72)):*
s/security settings/permissions/

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-:-K7TzUUfg5qFdYKDqMSi:676834233)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tz8aw3Ef8Z1ca0DkM:-K7Tz8ax98Q-OT7ek1Ew:-1787977015) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L43)):*
I think it's time to discuss where this is going and how the range logs will tie into the other event logs. I was envisioning that there would be a single event log table containing range events, cluster events, node events, table events, etc. It sounds like others have a different opinion and this change is moving in the direction of a log per event source (range, cluster, node, etc). The upside to this approach is more structured data per log, but it can lead to a proliferation of different logs that will have to be merged to display a unified log in the admin UI.

---

*[storage/log.go, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TzgGQ-4gJsUh_toiu:-K7TzgGRdrBVw_G8-uVl:207526734) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L44)):*
All of the cockroach tables belong in the `system` database. Pretty sure I made this comment in a previous PR. Happy to regenerate the rationale if requested.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 8 of 8 files at r1, 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Txvp8HVwxeARRIyQk:-K7VFotRH-jhIYFAQoMs:909054019) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L85)):*
is that check good for anything?

---

*[storage/log.go, line 103 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7VHlWR2NUE9fT7_o-i:-K7VHlWR2NUE9fT7_o-j:1279906528) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L103)):*
can't you drop that boolean and just not log if you have a `nil` `InternalExecutor`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7WsbDW5cgqhvI6TBI7:-K7WsbDW5cgqhvI6TBI8:1658368307) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L51)):*
How do you envision this table being used? Displaying splits by range ID doesn't seem particularly compelling from a UI perspective. As an admin I'd want to know the start and end keys. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[sql/internal.go, line 1 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tud6TcCcfZMjuFSX6:-K7Xhb9ayBa0LcMceUMJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/sql/internal.go#L1)):*
Done.

---

*[sql/internal.go, line 24 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TunEoTnVn3dyfet7v:-K7Xi4XAFvZxvvQWMuZ6:-1971348257) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/sql/internal.go#L24)):*
Reworded a bit.

---

*[sql/internal.go, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TuZOAD3ivw6z6qrd-:-K7Xi8yoonJRJ1T4dBTs:-2019671157) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/sql/internal.go#L34)):*
Didn't need to be, fixed.

---

*[storage/log.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TvpDk85_PUzyNsuMZ:-K7XnaMNz5pEv_9pwQJJ:309814830) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L33)):*
Any untyped constant can be silently coerced into another type; using integers instead of strings wont help with that, unless you mean that the "stringer" methods will panic for invalid values of the enumeration.

I also tend to prefer placing the strings into the database directly in order to make querying a better experience; it's easier to read the results when the enum value make sense.

---

*[storage/log.go, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TwMXDkZYyH66IZoSC:-K7XvbmUcX2FH04auuzL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L40)):*
Done.

---

*[storage/log.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tz8aw3Ef8Z1ca0DkM:-K7XvfZZCgkehTHVIU-i:-1660070481) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L43)):*
It's based on little more than thought experiment, but i'm convinced that a multi-log structure is the better course of action. Events that are rarer would likely contain more valuable information than "noisy" events like rebalances.  A combined table may also be wide and sparse, with many columns being unused for each event. We can obviously support such a table, but querying it might become inconvenient.

---

*[storage/log.go, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TzgGQ-4gJsUh_toiu:-K7X_fY2JX5o1bN2EXk_:1179887138) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L44)):*
I can't easily put this in the system table without significant refactoring, and I don't think that issue is settled. I'm currently of the opinion that we should have a "systemlog" database which contains the various log structured tables.

---

*[storage/log.go, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7WsbDW5cgqhvI6TBI7:-K7X_xDwnvbYMwdYxjWE:-1667038634) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L51)):*
Yes, I want to record the descriptor of the range at the time the split occurs.  However, that involves marshalling a proto to a bytes field, and was tricky enough that I held off on adding it as part of this PR (which has so much other stuff in it).

---

*[storage/log.go, line 71 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TweB9YwtfxB1V9ulj:-K7Y2Ya61YtOColNxmr0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L71)):*
Done.

---

*[storage/log.go, line 85 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Txvp8HVwxeARRIyQk:-K7Y3-Gf1OMMxQ6EjN_0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L85)):*
Done.

---

*[storage/log.go, line 100 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Ty7B_pu0qrH6p46l0:-K7Y40f1UGmCsI6bmPQH:-807793268) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L100)):*
I see "left" being used only in the context of Merges.

A difficulty with the nomenclature of these columns is the different event types; splits have "updated" and "new" range, merges have a "subsuming" and "subsumed" ranges, replica changes involve only one range.  If we're storing these events in a single table, we're going to want to re-use the same columns, so I chose a generic name of "otherRange" to hold information about the second range involved, if present.

This is all subject to change depending on the schema we eventually decide on, and there is still a lot of controversy about this matter; therefore, i'm going to leave this as-is for now, because it is subject to change after discussion.

---

*[storage/log.go, line 103 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7VHlWR2NUE9fT7_o-i:-K7Y3U5apUTHLgWn4B7F:2070563918) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L103)):*
Maybe, but this seems more clear. There's no guarantee that logs will be the only use of the SQLExecutor.

---

*[storage/log_test.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TyQjdcftnbrvSzAP8:-K7Y6J4J1E6O-uWfrmZg:-1674740189) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L48)):*
I thought I did add a TODO, good catch.

---

*[storage/log_test.go, line 77 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tyj_maPsk7ffy_2rZ:-K7Y8RjB2gp16mPTbRYM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L77)):*
Done.

---

*[storage/log_test.go, line 82 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tz5YMWGqioAXB5PKh:-K7Y8YRAA2N9hkTKs2V7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L82)):*
Done.

---

*[storage/log_test.go, line 88 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TywR1H3ISekX7Bckc:-K7Y8b2mI_SYxP6Fe_Mm:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L88)):*
Done.

---

*[storage/log_test.go, line 96 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tyz2t0PsmIWbC4dLQ:-K7Y8eajOW3LHfRp_G_M:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log_test.go#L96)):*
Done.

---

*[testutils/sqlutils/conn.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtZZ27ZFG9w72iQqJ:-K7YAWH5GktrpLR1FNzS:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/testutils/sqlutils/conn.go#L33)):*
Done.

---

*[testutils/sqlutils/conn.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtQFA9sLnhA2H4Xrq:-K7XbDFRI8DAX1I5uBpW:203458472) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/testutils/sqlutils/conn.go#L40)):*
Because I originally put this on the TestServer directly, but that was disallowed because securitytest is not allowed to be imported into the main product. So I then moved TestServer back into the parameters list, putting it at the front. It's an arbitrary decision, I can swap it back if you'd like.

---

*[util/testing.go, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Ttr7qK-d2j_NtOFhS:-K7Y9ryCxv29xH9eykeY:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/util/testing.go#L69)):*
Done.

---

*[util/testing.go, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtvbmdBUi-tKCQWyr:-K7Y9sIFs-7VjbikoKh-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/util/testing.go#L72)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 10 of 15 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tz8aw3Ef8Z1ca0DkM:-K7YBb_an0_PVFmIXuyO:1262202753) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L43)):*
Do you mind elaborating on the thought experiment? You look to be putting quite a bit of effort into making this log structured but it isn't clear to me what usage is going to want that structure. You've undoubtedly thought about this more than I have. Please educate me. 

---

*[storage/log.go, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TzgGQ-4gJsUh_toiu:-K7YBGx0L7pNNynn-js8:1553731871) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L44)):*
Eh? The `system` database is just a namespace. It doesn't affect whether this table is gossiped or anything else. I think it should be trivial to place it there. 

---

*[storage/log.go, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7WsbDW5cgqhvI6TBI7:-K7YCQU8beUgq7DFx7UU:-1926279933) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L51)):*
So what use is the range ID if you have the range descriptor (which will have the range ID inside of it)? Do you envision querying this table by range ID?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TvpDk85_PUzyNsuMZ:-K7YCpacVPmo_ZKpjEWk:1568907810) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L33)):*
stringer won't panic, but you'll get a string that clearly indicates it isn't a valid enum (something like "RangeEventLogType(5)"). The current approach allows you to accidentally insert garbage and never realize.

---

*[storage/log.go, line 71 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7YDOe8sWxpINEh5BCh:-K7YDOe9-6L4tg0uCVTJ:-184990716) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e786c526dd12e2f57a3eb6b035275b7af0b408c/storage/log.go#L71)):*
i could be wrong, but I thought an array here `[...]` instead of `[]` would avoid an allocation.

---

*[storage/log_test.go, line 49 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7YDW_FuTFG9vCPGGAi:-K7YDW_FuTFG9vCPGGAj:2006505037) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e786c526dd12e2f57a3eb6b035275b7af0b408c/storage/log_test.go#L49)):*
s/with/when/

---

*[testutils/sqlutils/conn.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtQFA9sLnhA2H4Xrq:-K7YDi3Rg1nq5LbLvVQg:-644711158) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/testutils/sqlutils/conn.go#L40)):*
I think the convention is that the `testing.T`-like thing is first in the parameter list, so I would like to see this switched back.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 71 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7YDOe8sWxpINEh5BCh:-K7YEkPtNToySQMY4nkm:-1205908303) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e786c526dd12e2f57a3eb6b035275b7af0b408c/storage/log.go#L71)):*
An array wouldn't allocate here, but it likely would cause an allocation when the arguments get passed to `ExecuteStatementInTransaction`. I can't imagine this is a performance concern given this code is only invoked on splits (not every operation).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7WsbDW5cgqhvI6TBI7:-K7budKGSPi3F0u22L4r:-2011776448) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L51)):*
I envision querying this table by range ID, although I think it's more for our use as developers than for external admins (by analogy with the similar per-tablet logging in bigtable). 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

Commit:
Adjust signature of sqlutils.PGUrl()

Move the util.Tester parameter to be the first parameter, in order to be more
idiomatic with other test functions.

mrtracy commented:
Review status: 4 of 15 files reviewed at latest revision, 23 unresolved discussions.

---

*[storage/log.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TvpDk85_PUzyNsuMZ:-K7mHcEbdwL3ZNdkR-TM:-209603342) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L33)):*
I'm going to take some time to write a (small) proposal on the structure of these log tables. Right now, I'm still personally leaning towards using strings here, because when querying this table from the command line the value will make sense immediately (without having to consult a lookup table). However, I want to consider this an open question for the time being and still go ahead with merging this particular PR (the structure of the table is only a small part of the PR)

---

*[storage/log.go, line 43 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7Tz8aw3Ef8Z1ca0DkM:-K7mI6wK7fHLvh0nzV8G:2067927451) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L43)):*
As mentioned in another comment (and in our 1-on-1 discussion), i'm going to prepare a small document analyzing the pros and cons of both approaches (multi-table vs. single table). However, i'd like to consider this as beyond the scope of this PR.

---

*[storage/log.go, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TzgGQ-4gJsUh_toiu:-K7mINis5WHolMRM6i8K:-44559063) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L44)):*
This is no longer relevant, as this table was moved into the `system` database in #3737

---

*[storage/log.go, line 51 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7WsbDW5cgqhvI6TBI7:-K7mIXiY4XyOGx-0N60x:641605813) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L51)):*
This will be part of a pros/cons document I will prepare on this topic. For now, it should be considered beyond the scope of this PR.

---

*[storage/log.go, line 71 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7YDOe8sWxpINEh5BCh:-K7mIfV7Bs3_2VVc8SV_:-73983003) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e786c526dd12e2f57a3eb6b035275b7af0b408c/storage/log.go#L71)):*
I'm going to assume that the current code is acceptable.

---

*[storage/log_test.go, line 49 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7YDW_FuTFG9vCPGGAi:-K7mLHbkOrV-YU4j9CSN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e786c526dd12e2f57a3eb6b035275b7af0b408c/storage/log_test.go#L49)):*
Done.

---

*[testutils/sqlutils/conn.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TtQFA9sLnhA2H4Xrq:-K7mODG_IjhLT1csfthp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0a488813507a4d0f71f18feeda4a00f57b47edc/testutils/sqlutils/conn.go#L40)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r4.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[storage/log.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TvpDk85_PUzyNsuMZ:-K7mUQP8vU5klKcK-aCE:-1375084444) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L33)):*
Let me clarify; I'm not suggesting you log integers to the table. I'm merely saying that while you're passing these constants around in memory, they should be integers that implement fmt.Stringer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: 8 of 15 files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[storage/log.go, line 33 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7rRI7lp4O3_v66f0dz:-K7rRI7lp4O3_v66f0e-:1690644240) ([raw file](https://github.com/cockroachdb/cockroach/blob/86963af77d55461b74729a2d3031a47083bea469/storage/log.go#L33)):*
Does this type need to be exported?

---

*[storage/log.go, line 37 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7rRFSkW-ShsArGWlD7:-K7rRFSkW-ShsArGWlD8:420446438) ([raw file](https://github.com/cockroachdb/cockroach/blob/86963af77d55461b74729a2d3031a47083bea469/storage/log.go#L37)):*
Does this constant need to be exported?

---

*[storage/log_test.go, line 50 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7rRR2NLTB44pchQsl3:-K7rRR2NLTB44pchQsl4:1424431470) ([raw file](https://github.com/cockroachdb/cockroach/blob/86963af77d55461b74729a2d3031a47083bea469/storage/log_test.go#L50)):*
I'll probably get flack for this, but formatting `RangeEventLogSplit` into the query is excessive, IMO. You can just put `eventType = 'split'` into the query. If the query doesn't match the event type for whatever reason your test will fail.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-:-K7rRtgqFJij6P7ahBLQ:676834233)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 7 files at r5.
Review status: 6 of 15 files reviewed at latest revision, 24 unresolved discussions.

---

*[storage/log.go, line 37 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7rRFSkW-ShsArGWlD7:-K7rSRfj5Bej8ZFwKyXw:2032887692) ([raw file](https://github.com/cockroachdb/cockroach/blob/86963af77d55461b74729a2d3031a47083bea469/storage/log.go#L37)):*
Well, it does because it's being interpolated into the query you commented on. I'd leave it exported, but that's me.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 9 of 15 files reviewed at latest revision, 24 unresolved discussions.

---

*[storage/log_test.go, line 50 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7rRR2NLTB44pchQsl3:-K7rVCUZIcAUXz3neb50:1546480387) ([raw file](https://github.com/cockroachdb/cockroach/blob/86963af77d55461b74729a2d3031a47083bea469/storage/log_test.go#L50)):*
I think it's definitely going to make sense once we flesh out this log system a bit more, with additional types and probably some query helpers. I'd like to keep those constants exported for the time being as well.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 9 of 15 files reviewed at latest revision, 23 unresolved discussions.

---

*[storage/log_test.go, line 50 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7rRR2NLTB44pchQsl3:-K7rW4HKM-MAtj15OIHV:-1983404302) ([raw file](https://github.com/cockroachdb/cockroach/blob/86963af77d55461b74729a2d3031a47083bea469/storage/log_test.go#L50)):*
Ok, though you're counting splits in this function. I'm not sure what else you'd want to pass in the query.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r6.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[storage/log.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TvpDk85_PUzyNsuMZ:-K7rWYyflpSfjp5j-wTy:-1383974699) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L33)):*
Looks like you "acknowledged" this comment, but it needs a response.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[storage/log.go, line 33 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687#-K7TvpDk85_PUzyNsuMZ:-K7rhItfqggJE2MI8mmM:-116687631) ([raw file](https://github.com/cockroachdb/cockroach/blob/30f7ad427c47aebd42a5c1b727bbda2117f5ccee/storage/log.go#L33)):*
I think it is fine as is, that this is a very low risk and not worth additional work in this PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3687)*
<!-- Sent from Reviewable.io -->
